core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -13,35 +13,252 @@
+    {
+        ensureBytesCapacity(currentOffset + newByteCount);
+
+        for (int i = 0; i < count; i++) {
+            int position = positions[i];
+            if (!block.isNull(position)) {
+                int length = lengths[i];
+                Slice slice = block.getSlice(position, 0, length);
+                slice.getBytes(0, bytes, targetOffsets[targetOffsetsIndex + i], length);
+            }
+        }
+
+        positionCount += count;
+        currentOffset += newByteCount;
+        updateSize(count, newByteCount);
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock block)
+    {
+        int rlePositionCount = block.getPositionCount();
+        if (rlePositionCount == 0) {
+            return;
+        }
+        int sourcePosition = 0;
+        ensurePositionCapacity(positionCount + rlePositionCount);
+        if (block.isNull(sourcePosition)) {
+            int offset = this.offsets[positionCount];
+            Arrays.fill(valueIsNull, positionCount, positionCount + rlePositionCount, true);
+            Arrays.fill(offsets, positionCount + 1, positionCount + rlePositionCount + 1, offset);
+            positionCount += rlePositionCount;
+
+            hasNullValue = true;
+            updateSize(rlePositionCount, 0);
+        }
+        else {
+            int startOffset = offsets[positionCount];
+            hasNonNullValue = true;
+            duplicateBytes(block.getValue(), sourcePosition, rlePositionCount, startOffset);
+        }
+    }
+
+    /**
+     * Copy {@code length} bytes from {@code block}, at position {@code position} to {@code count} consecutive positions in the {@link #bytes} array.
+     */
+    private void duplicateBytes(Block block, int position, int count, int startOffset)
+    {
+        int length = block.getSliceLength(position);
+        int newByteCount = count * length;
+        ensureBytesCapacity(currentOffset + newByteCount);
+
+        for (int i = 0; i < count; i++) {
+            Slice slice = block.getSlice(position, 0, length);
+            slice.getBytes(0, bytes, startOffset + (i * length), length);
+            this.offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
+        }
+
+        positionCount += count;
+        currentOffset += newByteCount;
+        updateSize(count, newByteCount);
+    }
+
+    @Override
+    public Block build()
+    {
+        if (!hasNonNullValue) {
+            return new RunLengthEncodedBlock(NULL_VALUE_BLOCK, positionCount);
+        }
+        VariableWidthBlock result = new VariableWidthBlock(
+                positionCount,
+                Slices.wrappedBuffer(bytes, 0, currentOffset),
+                offsets,
+                hasNullValue ? Optional.of(valueIsNull) : Optional.empty());
+        reset();
+        return result;
+    }
+
+    private void reset()
+    {
+        initialEntryCount = calculateNewArraySize(positionCount);
+        initialized = false;
+        valueIsNull = new boolean[0];
```

Comment:
I think we miss something line `initialBytesArraySize`. `VariableWidthBlockBuilder` uses `calculateBlockResetBytes`